### DISCUSSION TITLE
Render is very slow

### DISCUSSION THREAD

**Discussion by fabricelallemand15** [quarto-dev/quarto-cli#4684] [Q&A]
Hi
I'm using quarto to build a website. Each page is rendered to HTML and to PDF.
Is there a way to update the website by rendering only the new or the modified files, without considering files which did not change?
Thanks

**Comment by cderv** [comment]
Did you already have a look at the `freeze` option and what incremental render is ? 
Doc is here for details: https://quarto.org/docs/projects/code-execution.htm

By freezing content, it will speed rendering by not recomputing document with computational content. 

If the slow part is the PDF rendering of document, it may not speed up that much. 

If you have a project to share, we are interested. Improving performance is important for us !

**Comment by fabricelallemand15** [comment]
Yes, I had a look at the freeze option.
The slow part in my case is the PDF rendering. I think I didn't really understand what 'incremental render' is, and I should probably render only my new files to update my website.
Thanks

**Reply by cderv** [reply]
> I think I didn't really understand what 'incremental render' is

By passing a file or a folder to `quarto render`, you'll be able to render a subset of your file but still have a book output for example. 
This works best for HTML version of work for example but we do our best for other format too. 

> The slow part in my case is the PDF rendering.

About that, this is tricky as it is not sure what quarto could do. Rendering to PDF is done directly by LaTeX that we execute but do no control specifically. 

> I'm using quarto to build a website. Each page is rendered to HTML and to PDF.

So you build a PDF file for each page of the website ? And this is what is re-run each time even if the source is not updated ? 
So you're building a new PDF whereas you don't need to ? Am I understanding correctly ?

For more custom re-rendering, you could maybe leverage [Project scripts](https://quarto.org/docs/projects/scripts.html) and [rendering target](https://quarto.org/docs/projects/quarto-projects.html#rendering-projects) to maybe tweak rendering of PDF phase in a second phase after HTML rendering. 

Curious to know more about the use case - maybe we can add something more in the rendering of projects for PDF for example if we detect source file has not change, and knowing will help understand

> Yes, I had a look at the freeze option.

Freeze option should help you save time on computation. Could be good to activate for your website, this should help making process more smooth

**Reply by memeplex** [reply]
Regarding this, I'm confused about what incremental rendering and render targets can do for pdf projects.

For example, this is my project config:

```
project:
  type: book

book:
  title: "mybook"
  chapters:
    - index.qmd
    - chapter1.qmd
    - chapter2.qmd

format:
  pdf:
    keep-tex: true
```

If I run `quarto render chapter1.qmd --to pdf` both chapters are included in the generated tex file.

If I add:

```
project:
  ...
  render:
    - chapter1.qmd
 ```

to the config and run `quarto render --to pdf` again both chapters are included.

The only way I found to work on a chapter by chapter basis is to comment out all other documents in the chapters section.

**Reply by mcanouil** [reply]
@memeplex please open a new thread or search for thread related to your "issue". I'm afraid this thread is not related to render targets or books. Thank you.
For instance:
- https://github.com/quarto-dev/quarto-cli/discussions/7762

**Reply by memeplex** [reply]
Ok, I asked here because the thread was about building a pdf and the suggestion was to look at rendering targets, but I'm going to open a new discussion for my "issue". Thanks.

**Comment by produnis** [comment]
I've encountered a similar issue... Currently, my blog comprises 376 documents, and I've configured the `freeze` option to cache computations, although there aren't many of them. Previously, when using `hugo`, rendering the site would take roughly 1 second. However, since transitioning to Quarto, each document takes approximately 1-2 seconds to render. Consequently, I find myself waiting around 400 seconds for my webpage to fully render.

I'm wondering if there's an alternative approach to cache all my existing entries so that Quarto doesn't need to render a new HTML from scratch each time. Any insights or suggestions would be greatly appreciated!

(my quarto website is <https://www.produnis.de>)

**Reply by danmackinlay** [reply]
Ausgezeichnetes website, @produnis .

Yes, I have about 1300 pages and I have come to the conclusion that [quarto websites are just slow](https://danmackinlay.name/notebook/quarto#this-is-slow); Typically it's >10 minutes to build my site, compared to a few seconds for hugo. I think this is a slightly different problem than the OP, though, who is mostly concerned about slow PDF build times

**Reply by cscheid** [reply]
@danmackinlay Thank you for the comprehensive and frank writeup: us in the Quarto team have read it and appreciate it. I understand the performance frustration especially.

We've been making slow-but-steady performance improvements. By our measurements, Quarto 1.4 is 25% faster than 1.2. That's cold comfort when your render time is still in the order of many minutes. We have a long-term plan to improve that we're implementing.

In generalities, our runtime is roughly spent 2/3s inside Pandoc, and 1/3 in Deno. Our Pandoc filtering infrastructure is pretty extensive, and some of our early decisions there have performance consequences that we didn't foresee: we're now working on them. In Deno, our performance profile is relatively flat, and so the work is going to be more of the "continued small fixes" kind.

Thank you again for the writeup, it's definitely appreciated.

**Reply by danmackinlay** [reply]
That's very informative @cscheid  thanks! I'll update the post with your information.

Yes I hope my post doesn't come across as whinging! To be clear, I no-one is forcing me to use quarto; I do because I think it is a good product and net positive for my workflow, and I'm very appreciative of the hard work that has gone into it. It's good to be aware of the frictions, though.

**Comment by alankjackson** [comment]
I have started using Leaflet for maps in Quarto, because it is literally 500 times faster to build over tmap where there are many polygons.
